src/video/kmsdrm/SDL_kmsdrmvideo.c

```diff
@@ -55,129 +55,106 @@
 #define KMSDRM_DRI_PATH openbsd69orgreater ? "/dev/dri/" : "/dev/"
-#define KMSDRM_DRI_DEVFMT openbsd69orgreater ? "%scard%d" : "%sdrm%d"
 #define KMSDRM_DRI_DEVNAME openbsd69orgreater ? "card" : "drm"
-#define KMSDRM_DRI_DEVNAMESIZE openbsd69orgreater ? 4 : 3
-#define KMSDRM_DRI_CARDPATHFMT openbsd69orgreater ? "/dev/dri/card%d" : "/dev/drm%d"
 #else
 #define KMSDRM_DRI_PATH "/dev/dri/"
-#define KMSDRM_DRI_DEVFMT "%scard%d"
 #define KMSDRM_DRI_DEVNAME "card"
-#define KMSDRM_DRI_DEVNAMESIZE 4
-#define KMSDRM_DRI_CARDPATHFMT "/dev/dri/card%d"
 #endif
 
+#define KMSDRM_DRI_CARDPATHFMT KMSDRM_DRI_PATH KMSDRM_DRI_DEVNAME "%d"
+#define KMSDRM_DRI_DEVNAMESIZE sizeof(KMSDRM_DRI_DEVNAME) - 1
+#define KMSDRM_DRI_PATHSIZE    sizeof(KMSDRM_DRI_PATH) - 1
+
 #ifndef EGL_PLATFORM_GBM_MESA
 #define EGL_PLATFORM_GBM_MESA 0x31D7
 #endif
 
 static int
-check_modestting(int devindex)
+get_driindex(void)
 {
-    SDL_bool available = SDL_FALSE;
-    char device[512];
+    int available = -ENOENT;
+    char device[512] = KMSDRM_DRI_PATH;
     int drm_fd;
     int i;
-
-    SDL_snprintf(device, sizeof (device), KMSDRM_DRI_DEVFMT, KMSDRM_DRI_PATH, devindex);
-
-    drm_fd = open(device, O_RDWR | O_CLOEXEC);
-    if (drm_fd >= 0) {
-        if (SDL_KMSDRM_LoadSymbols()) {
-            drmModeRes *resources = KMSDRM_drmModeGetResources(drm_fd);
-            if (resources) {
-                SDL_LogDebug(SDL_LOG_CATEGORY_VIDEO,
-                  KMSDRM_DRI_DEVFMT
-                  " connector, encoder and CRTC counts are: %d %d %d",
-                  KMSDRM_DRI_PATH, devindex,
-                  resources->count_connectors, resources->count_encoders,
-                  resources->count_crtcs);
-
-                if (resources->count_connectors > 0
-                 && resources->count_encoders > 0
-                 && resources->count_crtcs > 0)
-                {
-                    available = SDL_FALSE;
-                    for (i = 0; i < resources->count_connectors; i++) {
-                        drmModeConnector *conn = KMSDRM_drmModeGetConnector(drm_fd,
-                            resources->connectors[i]);
-
-                        if (!conn) {
-                            continue;
-                        }
-
-                        if (conn->connection == DRM_MODE_CONNECTED && conn->count_modes) {
-                            if (SDL_GetHintBoolean(SDL_HINT_KMSDRM_REQUIRE_DRM_MASTER, SDL_TRUE)) {
-                                /* Skip this device if we can't obtain DRM master */
-                                KMSDRM_drmSetMaster(drm_fd);
-                                if (KMSDRM_drmAuthMagic(drm_fd, 0) == -EACCES) {
+    int devindex = -1;
+    DIR *folder = opendir(device);
+    if (!folder) {
+        SDL_SetError("Failed to open directory '%s'.\n", device);
+        return -ENOENT;
+    }
+
+    memcpy(device + KMSDRM_DRI_PATHSIZE, KMSDRM_DRI_DEVNAME,
+           KMSDRM_DRI_DEVNAMESIZE);
+    for (struct dirent *res; (res = readdir(folder));) {
```

Comment:
You may want to sort these, as you could get then in order card3, card1, card2, etc. and my understanding is that they are named in order of preference (or being attached to the system?)

Reply:
scandir() is probably the function you want here.

Reply:
There isn't really an order, it could be when you physically attach and detach, it could be when you load and unload a kernel module. In recent kernels it's commonplace because simpledrm driver registers with the kernel, then the hardware accelerated graphics driver registers. This causes the software graphics (simpledrm) to be unregistered, leaving you with card1 and no card0. So it could be multiple hardware or software reasons. What caused this bug is a misunderstanding that the order actually signified something.